[webkitcorepy] Fix race condition in FileLock

JonWBedard opened the pull request:
#### 1c10cac855587753cfb42e0e7f3b62e2480eb62e
<pre>
[webkitcorepy] Fix race condition in FileLock
<a href="https://bugs.webkit.org/show_bug.cgi?id=256931">https://bugs.webkit.org/show_bug.cgi?id=256931</a>
rdar://109487762

Reviewed by Alexey Proskuryakov.

* Tools/Scripts/libraries/webkitcorepy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/file_lock.py:
(FileLock.is_process_running): Check if a specific PID is running.
(FileLock.acquire): Create and acquire lock in a single operation on Unix systems.
Save the current PID to the file, which allows future processes to ignore this lock
if the PID saved in it is dead. For systems which support it, use O_EXLOCK and interrupts
instead of a spinlock. Raise exceptions when we fail to acquire the lock.
(FileLock.release): Delete the lockfile before releasing it on Unix systems. Systems using
O_EXLOCK simply need to close their file handle.
(FileLock.__exit__): Don&apos;t raise a 2nd exception if __enter__ has raised one.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/file_lock_unittest.py:
(action):
(FileLockTestCase.test_locked):
(FileLockTestCase.test_locked_timeout):
(FileLockTestCase.test_double):
(FileLockTestCase.test_race):

Canonical link: <a href="https://commits.webkit.org/264321@main">https://commits.webkit.org/264321@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6215bffe968c39b07f6d328fafc7c6ad4a8f7ddd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7413 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/7674 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/7847 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9045 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/7618 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7423 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/9584 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/7597 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/10503 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/7543 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/8645 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/6865 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9154 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/7527 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/5955 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/6744 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/14471 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/7211 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/6858 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10145 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7349 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6008 "Passed tests") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/7386 "Passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/6698 "Built successfully") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/1743 "Build was cancelled. Recent messages:OS: Monterey (12.6.3), Xcode: 13.4.1; Checked out pull request; Reviewed by Alexey Proskuryakov; Compiled WebKit (cancelled)") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/10906 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/858 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7084 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/file_lock.py:
```diff
@@ -22,60 +22,86 @@
+    @classmethod
+    def is_process_running(cls, pid):
+        try:
+            os.kill(pid, 0)
```
Seems worth asserting that this isn't getting called on Windows. The current code doesn't, but that's fragile.

aproskuryakov commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/file_lock.py:
```diff
@@ -22,60 +22,86 @@
+            os.kill(pid, 0)
+            return True
+        except OSError:
+            return False
```
Maybe check if the error is `errno.ESRCH`. I'm a little concerned about some of the autoinstallers running under `sudo`. But not an issue when building.

JonWBedard replied:
We already handle auto installers running under `sudo`, actually, we make sure that the entire directory is owned by the user after we finish installing a package. That will include this file. It's not a perfect system (if we manage to have Python killed when doing an install, pretty sure we'd leave some `sudo` files behind), but it's going to be pretty effective.

aproskuryakov commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/file_lock.py:
```diff
@@ -22,60 +22,86 @@
 
     def acquire(self):
         if self._descriptor:
+            self._depth += 1
```
Not super happy about making this a recursive lock, as that's generally an anti-pattern. Is this strongly needed?

aproskuryakov commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/file_lock.py:
```diff
@@ -22,60 +22,86 @@
                 time.sleep(0.01)
 
     def release(self):
+        self._depth -= 1
```
Let's assert that it's bigger than 0 first.

aproskuryakov commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/file_lock.py:
```diff
@@ -22,60 +22,86 @@
 
     def release(self):
+        self._depth -= 1
+        if self._depth or not self._descriptor:
```
And lacking `self._descriptor` should be fatal too.

JonWBedard replied:
I'll make it fatal here, but it shouldn't be fatal in `__exit__`.

One of the most common patterns in Python with a class like this one (and the way this class is mostly used) is as a context manager. What's notable about this is that `__exit__` will be called even if `__enter__` raises an exception, and in that case, `self._descriptor` will be `None`. We don't want to raise the 2nd exception in that case.

aproskuryakov replied:
I think that we want to terminate the process if it raises an exception when locking (as opposed to returning False). That's not a recoverable condition.

aproskuryakov commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/file_lock.py:
```diff
@@ -22,60 +22,86 @@
+            elif not self.USE_EXLOCK:
+                os.unlink(self.path)
         except (IOError, OSError):
             pass
```
I think that this is also better to make fatal.

aproskuryakov commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/file_lock.py:
```diff
@@ -22,60 +22,86 @@
+            with open(self.path) as file:
+                pid = file.readline().strip()
+            if self.INTEGER_RE.match(pid) and not self.is_process_running(int(pid)):
+                os.unlink(self.path)
```
There is a race here. If two processes get to this line, one can then get all the way to line 79 and create a new lock file, and then another process will delete it. I don't know if/how this is normally solved.

JonWBedard replied:
That's true, although I'm comfortable leaving this race condition in place, given I'm not familiar with the tools other systems we might be running on have access to.

It's worth pointing out that this race condition requires 2 things. The first is that a previous process was actually killed, because if a process is merely interrupted, we clean everything up as we unwind the stack. The second is that we have two processes then racing the initial lock at the same time. It seems like our other option would be to avoid the race condition by refusing to acquire the lock because we can't check if the process which originally acquired it is live. That seems strictly worse, because we still fail in your scenario, but now we also fail if we only have one process running after a previous process has been killed.

aproskuryakov commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/file_lock.py:
```diff
@@ -22,60 +22,86 @@
+                        os.close(self._descriptor)
                     self._descriptor = None
                     return False
                 time.sleep(0.01)
```
Not very happy about spinning, but I guess there is no other way if we want to have a timeout for locking. Maybe we don't have to, though?

JonWBedard replied:
There _is_ another way for `O_EXLOCK` (although it doesn't work for Windows and ports without `O_EXLOCK`, so we have to keep the spinning). Updating PR with the approach that doesn't use a spin lock.

aproskuryakov commented on Tools/Scripts/libraries/webkitcorepy/webkitcorepy/file_lock.py:
```diff
@@ -44,42 +62,60 @@ def acquired(self):
+                os.unlink(self.path)
+
+        if self.USE_EXLOCK and self.timeout:
+            with Timeout(
```
I think that this has a race. The alarm can fire after the file is locked by the kernel but before `self._descriptor` is assigned.

JonWBedard replied:
Is does, yes. But this exception is always fatal in practice, so we'll end up with a dead process shortly after throwing it.

JonWBedard committed:
[webkitcorepy] Fix race condition in FileLock
https://bugs.webkit.org/show_bug.cgi?id=256931
rdar://109487762

Reviewed by Alexey Proskuryakov.

* Tools/Scripts/libraries/webkitcorepy/setup.py: Bump version.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/__init__.py: Ditto.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/file_lock.py:
(FileLock.is_process_running): Check if a specific PID is running.
(FileLock.acquire): Create and acquire lock in a single operation on Unix systems.
Save the current PID to the file, which allows future processes to ignore this lock
if the PID saved in it is dead. For systems which support it, use O_EXLOCK and interrupts
instead of a spinlock. Raise exceptions when we fail to acquire the lock.
(FileLock.release): Delete the lockfile before releasing it on Unix systems. Systems using
O_EXLOCK simply need to close their file handle.
(FileLock.__exit__): Don't raise a 2nd exception if __enter__ has raised one.
* Tools/Scripts/libraries/webkitcorepy/webkitcorepy/tests/file_lock_unittest.py:
(action):
(FileLockTestCase.test_locked):
(FileLockTestCase.test_locked_timeout):
(FileLockTestCase.test_double):
(FileLockTestCase.test_race):

Canonical link: https://commits.webkit.org/264321@main
